Update configuration_metadata.go

AamnaZahid opened the pull request:
Support added for the configuration-properties.names property.

<!-- Thanks for contributing. To speed up the process of reviewing your pull
request please provide us with the following information: -->

## Summary
<!-- A short explanation of the proposed change -->

## Use Cases
<!-- An explanation of the use cases your change enables -->

## Checklist
<!-- Please confirm the following -->
* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [x] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [x] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [x] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

anthonydahanne commented:
Hello @AamnaZahid !
Thanks for your contribution!
Can you please:
* provide context / explain what those changes will enable?
* add unit tests

So that we'll be able to review.
Thank you!

AamnaZahid commented:
**Context/Explanation of Changes**

**Before:**
The function only processed the configuration-properties.classes field by splitting its values and appending them to the classes list. It then matched these classes against SourceType values in metadata groups and properties.

**After:**
In addition to classes, code now handle configuration-properties.names similarly, where the names values are also processed and stored in a names list.
Both classes and names lists are merged into a combined list, which is used to match against the SourceType or Name fields in metadata Groups and Properties.
This ensures that both properties (classes and names) are additive and provide more flexibility for users to define mappings based on either class or name.

**What those changes will enable?**

1. _Support for configuration-properties.names:_  Now, users can specify both classes and names in the dataflow-configuration-metadata.properties file. This makes the configuration process more intuitive and documented.
2. 
3. _Expanded Mapping Capabilities:_ The new logic allows for mapping based on both SourceType and Name fields, rather than being restricted to just SourceType. This provides more flexibility in defining and processing configuration metadata.

dmikusa commented:
> Support for configuration-properties.names: Now, users can specify both classes and names in the dataflow-configuration-metadata.properties file. This makes the configuration process more intuitive and documented.

OK, interesting. This is what I was trying to figure out as well. Where do these files come from, what apps are depending on them. 

Can you provide a doc link that explains these files a little more? I want to make sure that there is some context here so we, as maintainers, know how this feature is being used. That'll prevent regressions in the future. Thanks

anthonydahanne commented:
is that enough @dmikusa ?
https://dataflow.spring.io/docs/applications/application-metadata/

AamnaZahid commented:
Yes

On Wed, 23 Oct 2024, 07:22 Anthony Dahanne, ***@***.***>
wrote:

> is that enough @dmikusa <https://github.com/dmikusa> ?
> https://dataflow.spring.io/docs/applications/application-metadata/
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/paketo-buildpacks/spring-boot/pull/513#issuecomment-2430688216>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A324ALWFDQAOGPUCBZ4CVQDZ44B7FAVCNFSM6AAAAABPHAMSK2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMZQGY4DQMRRGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

anthonydahanne commented:
@AamnaZahid I fixed a test, and added another one in a separate commit.

AamnaZahid commented:
Awesome, thanks for the update!

On Thu, 24 Oct 2024, 04:40 Anthony Dahanne, ***@***.***>
wrote:

> @AamnaZahid <https://github.com/AamnaZahid> I fixed a test, and added
> another one in a separate commit.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/paketo-buildpacks/spring-boot/pull/513#issuecomment-2433787014>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A324ALRRZ7EIIP5LRQ2G3KTZ5AXWVAVCNFSM6AAAAABPHAMSK2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMZTG44DOMBRGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

onobc commented:
To elaborate a bit more on the link provided by @anthonydahanne .

Here is a little background on the Dataflow dedicated metadata. It is used to drive the UI/API during stream/task definition. 

This describes why we have the dedicated metadata:
https://dataflow.spring.io/docs/applications/application-metadata/#dedicated-metadata-artifacts

This describes the metadata file:
https://dataflow.spring.io/docs/applications/application-metadata/#data-flow-configuration-metadata

This describes the metadata label:
https://dataflow.spring.io/docs/applications/application-metadata/#metadata-container-image-label

Here is a concrete example that uses 3 of the pre-built stream-applications ([e.g. splitter used below](https://docs.spring.io/stream-applications/docs/current/reference/html/index.html#spring-cloud-stream-modules-splitter-processor) that include said metadata.

The stream is defined as `time | splitter | log` and you can see in the UI the user is presented w/ only the properties that are in this resulting metadata (otherwise we would have to show them all N of the available Spring Boot config props).


<img width="821" alt="Screenshot 2024-10-24 at 09 37 50" src="https://github.com/user-attachments/assets/cee770a9-e90d-46db-abb4-5f215e11525c">

<img width="806" alt="Screenshot 2024-10-24 at 09 37 58" src="https://github.com/user-attachments/assets/3688a760-759e-498c-81e2-bdd216961e6e">

I hope this helps w/ some more context.

Thanks
